File: examples/python/graph_1.py
from collections import deque


def bfs(graph, start):
    visited = set()
    q = deque([start])
    result = []
    while q:
        v = q.popleft()
        if v in visited:
            continue
        visited.add(v)
        result.append(v)
        for u in graph.get(v, []):
            if u not in visited:
                q.append(u)
    return result


if __name__ == "__main__":
    g = {0: [1, 2], 1: [2], 2: [0, 3], 3: [3]}
    print(bfs(g, 2))

File: examples/rust/graph_1.rs
use std::collections::VecDeque;

fn bfs(graph: &[Vec<usize>], start: usize) -> Vec<usize> {
    let mut visited = vec![false; graph.len()];
    let mut q = VecDeque::from([start]);
    let mut result = Vec::new();
    while let Some(v) = q.pop_front() {
        if visited[v] {
            continue;
        }
        visited[v] = true;
        result.push(v);
        for &u in &graph[v] {
            if !visited[u] {
                q.push_back(u);
            }
        }
    }
    result
}

fn main() {
    let g: Vec<Vec<usize>> = vec![vec![1, 2], vec![2], vec![0, 3], vec![3]];
    println!("{:?}", bfs(&g, 2));
}
